File: petties-web/src/services/reportService.ts
/**
 * Report API Service
 */
import axios from './api/client';
import type { ResolveReportRequest, ReportResponse, ReportStatus } from '../types/report';

// Spring Page response type
interface PageResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

const REPORT_API = '/reports';
const ADMIN_REPORT_API = '/admin/reports';

/**
 * Tạo báo cáo (multipart): BE upload ảnh lên Cloudinary.
 */
export const createReport = async (
    bookingId: string,
    reason: string,
    imageFiles: File[]
): Promise<ReportResponse> => {
    const formData = new FormData();
    formData.append('bookingId', bookingId);
    formData.append('reason', reason);
    for (const file of imageFiles) {
        formData.append('files', file);
    }
    const response = await axios.post<ReportResponse>(REPORT_API, formData);
    return response.data;
};

/**
 * Get current user's reports
 */
export const getMyReports = async (
    page: number = 0,
    size: number = 20
): Promise<PageResponse<ReportResponse>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());

    const response = await axios.get(`${REPORT_API}/my?${params.toString()}`);
    return response.data;
};

/**
 * Get reports by clinic ID (Admin, Clinic Owner, Clinic Manager)
 */
export const getClinicReports = async (
    clinicId: string,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<ReportResponse>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());

    const response = await axios.get(`${REPORT_API}/clinics/${clinicId}?${params.toString()}`);
    return response.data;
};

/**
 * Get all reports (Admin only)
 */
export const getAllReportsForAdmin = async (
    status?: ReportStatus,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<ReportResponse>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    if (status) params.append('status', status);

    const response = await axios.get(`${ADMIN_REPORT_API}?${params.toString()}`);
    return response.data;
};

/**
 * Resolve report (Admin only)
 */
export const resolveReport = async (
    reportId: string,
    request: ResolveReportRequest
): Promise<ReportResponse> => {
    const response = await axios.put(`${ADMIN_REPORT_API}/${reportId}/resolve`, request);
    return response.data;
};

/**
 * Cập nhật báo cáo PENDING (multipart): ảnh giữ lại trong `existingAttachmentUrlsJson`, file mới trong `files`.
 * Dùng POST .../update (PUT + multipart thường bị client/proxy làm mất body).
 */
export const updateReport = async (
    reportId: string,
    reason: string,
    newImageFiles: File[],
    keptAttachmentUrls: string[]
): Promise<ReportResponse> => {
    const formData = new FormData();
    formData.append('reason', reason);
    formData.append('existingAttachmentUrlsJson', JSON.stringify(keptAttachmentUrls));
    for (const file of newImageFiles) {
        formData.append('files', file);
    }
    const response = await axios.post<ReportResponse>(`${REPORT_API}/${reportId}/update`, formData);
    return response.data;
};

/** Withdraw own pending report (sets status WITHDRAWN) */
export const withdrawReport = async (reportId: string): Promise<ReportResponse> => {
    const response = await axios.delete(`${REPORT_API}/${reportId}`);
    return response.data;
};

// Named export for backwards compatibility
export const reportService = {
    createReport,
    updateReport,
    withdrawReport,
    getMyReports,
    getClinicReports,
    getAllReportsForAdmin,
    resolveReport,
};

File: petties_mobile/lib/data/services/report_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import '../models/report.dart';
import 'api_client.dart';

class ReportService {
  final ApiClient _apiClient = ApiClient.instance;

  Future<void> createReport({
    required String bookingId,
    required String reason,
    List<File>? imageFiles,
  }) async {
    final formData = FormData.fromMap({
      'bookingId': bookingId,
      'reason': reason,
    });

    if (imageFiles != null && imageFiles.isNotEmpty) {
      for (var file in imageFiles) {
        formData.files.add(MapEntry(
          'files',
          await MultipartFile.fromFile(file.path, filename: file.path.split('/').last),
        ));
      }
    }

    await _apiClient.post(
      '/reports',
      data: formData,
      options: Options(contentType: 'multipart/form-data'),
    );
  }

  Future<void> updateReport({
    required String reportId,
    required String bookingId,
    required String reason,
    List<File>? imageFiles,
  }) async {
    final formData = FormData.fromMap({
      'bookingId': bookingId,
      'reason': reason,
    });

    if (imageFiles != null && imageFiles.isNotEmpty) {
      for (var file in imageFiles) {
        formData.files.add(MapEntry(
          'files',
          await MultipartFile.fromFile(file.path, filename: file.path.split('/').last),
        ));
      }
    }

    await _apiClient.put(
      '/reports/$reportId',
      data: formData,
      options: Options(contentType: 'multipart/form-data'),
    );
  }

  Future<void> deleteReport(String reportId) async {
    await _apiClient.delete('/reports/$reportId');
  }

  Future<List<ReportResponse>> getMyReports({int page = 0, int size = 20}) async {
    final response = await _apiClient.get(
      '/reports/my',
      queryParameters: {
        'page': page,
        'size': size,
      },
    );

    if (response.data != null && response.data['content'] != null) {
      final List content = response.data['content'];
      return content.map((e) => ReportResponse.fromJson(e)).toList();
    }
    return [];
  }
}

